[singlehtml] add docname to section anchor to make them unique

gastmaier opened the pull request:
## Purpose

Follow up to #13717, inverting the logic, instead of patching the toctree to yield "#id1" instead of "#document-path/to#id1", have the section id to be docname preffixed, solving non-unique ids in singlehtml.
Allows to remove post Sphinx transforms like in [here](https://github.com/analogdevicesinc/doctools/blob/main/adi_doctools/__init__.py#L131-L173)

### Top level overview of current behavior

* ID collision is resolved per doc (#already-used -> #id1,  #already-used -> #id2).
* There is no ID collision resolution on singlehtml step.

### Approach taken

Based on the LaTeX builder solution.
`sphinx/writers/latex.py#hypertarget[withdoc=True]` method suffixes docutils id with the docname.
In my implementation I edit ids['0'] directly to not have to overwrite the whole `visit_section` method, but I understand if requested to not modify the tree and instead overwrite.

### On the format #document-test/extra#id1

It is compatible with HTML anchoring, CSS and JavaScript selectors, but require escaping:
```css
#document-test\/extra\#test {color: #f00;}
```

```javascript
document.querySelector('#document-test\\/extra\\#test')
```

### Tests

The following tests are relevant:

* tests/test_builders/test_build_html_tocdepth.py
* test_build_html_numfig.py


## References

- #13717 (replaces)
- #11970,
- #12551
- #13037

akhilsmokie7-cloud reviewed (commented):
Align with purpose

gastmaier committed:
Add test to assert unique ids in singlehtml output

Since the singlehtml aggregates all doc files into a single html page
during the write step, and the ids must be unique for proper link
anchoring, add test that collects all ids in the page and checks if all
ids are unique, by asserting the length of the list against it as a set.

gastmaier committed:
Add section with same title for test-tocdepth/[foo/bar]

To assert unique ids in singlehtml builder.

Signed-off-by: Jorge Marques <jorge.marques@analog.com>

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -110,7 +110,7 @@ def assemble_toc_secnumbers(self) -> dict[str, dict[str, tuple[int, ...]]]:
         for docname, secnums in self.env.toc_secnumbers.items():
             for id, secnum in secnums.items():
-                alias = f'{docname}/{id}'
+                alias = f'{docname}{id}'
```
What kind of values are possible for the `docname` and `id`?

jayaddison replied:
(also: I guess people _shouldn't_ have written hyperlinks or saved bookmarks with the assumption that these aliases are stable?  but, even so - if we change the format, I guess we would break those?)

jayaddison replied:
@gastmaier in fact: I'm not sure where these `/` separator characters appear.  What does this code relate to?

gastmaier replied:
For singlehtml and at the assemble toctree step, the href is a tuple of docname and refid.
`#document-path/to/#id1` to  try to avoid the refid confliction in singlehtml mode problem, which didn't work because it would patch toctree, but the content body still had the non-unique ids.

My pr changes the toctree href format from
`#document-path/to/#id1`  to `#document-path/to#id1` (removes end slash)
and for content ids from
`#d1`  to `#document-path/to#id1` (adds doc prefix to make unique)
the new template is therefore:
 `#document-{doc}#{id}`
direct tuple of docname and refid, without the slash.

These are valid HTML anchors, but do require escaping when manipulating with:
css
```
#document-test\/extra\#test {color: #f00;}
```
and javascript
```
document.querySelector('#document-test\\/extra#test')
```

gastmaier replied:
[singlehtml.zip](https://github.com/user-attachments/files/22092103/singlehtml.zip)
here is a singlehtml build with the patch

jayaddison commented on sphinx/writers/html5.py:
```diff
@@ -497,6 +498,15 @@ def depart_term(self, node: Element) -> None:
             self.body.append('</dt>')
 
+    def visit_section(self, node: section) -> None:
+        if self.builder.name == 'singlehtml' and node['ids']:
```
We don't seem to use many `@property` methods in the Sphinx writers, but maybe this `singlehtml` condition is getting to the point where it makes sense (this is the third potential callsite, I think?).

gastmaier committed:
[singlehtml] Append docname to refid and ids

Use doc path to make ids unique with format ``/docname/#id``.
Compensates for the loss of the pathname in the href.
This will break existing hyperlinks to `singlehtml` HTML documents since
it alters the format

gastmaier committed:
Update AUTHORS.rst and CHANGES.rst

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -88,29 +88,47 @@ def _get_local_toctree(
+            if 'refid' in node:
+                node['refid'] = 'document-' + docname + '#' + node['refid']
+            if 'ids' in node:
+                node['ids'] = ['document-' + docname + '#' + id for id in node['ids']]
```
I'll plan to do this within the next 24h or so, but I'll ask in case it is something you could do quickly: could you print out two columns of text with the `before` and `after` values for these node attributes when building a non-trivial project (easiest/safest choice: Sphinx itself)?

e.g.

```
refids
before               after
[sample/#foo]        [document-sample#foo]

node_id
sample/#foo          document-sample#foo
```

The reason I ask: I'd like to inspect the places where the results differ, and in particular how the code changes achieve uniqueness of the results.

jayaddison replied:
(I'm also wondering whether `docutils` -- which produces the `node` objects, if I understand correctly - could help us and allow us to fix this in a more central location; and I hope that viewing the comparison columns may also help to understand whether that is realistic or whether this is some Sphinx-specific quirk)

jayaddison replied:
> (I'm also wondering whether `docutils` -- which produces the `node` objects, if I understand correctly - could help us and allow us to fix this in a more central location; and I hope that viewing the comparison columns may also help to understand whether that is realistic or whether this is some Sphinx-specific quirk)

Nope, scratch that - I think that `docutils` is unaware of the notion of `docnames`, so whatever is going on here must, I think, be part of Sphinx itself.

gastmaier replied:
So docutils provides the solved tree to the builder, with each doc being a document.
Sphinx guarantees the ids are unique per doc, the filesystem guarantees the docname is unique (you cannot have two identical paths)
But the builder singlehtml flattens all into the root doc `index`, loosing the information of the docname, causing non-unique ids after flatting it.
This fix recovers the docname and patches into the id itself.

The sphinx documentation itself, attached below, has conflicts, there are many duplicated id1.

[singlehtml.zip](https://github.com/user-attachments/files/22115511/singlehtml.zip)

The table requested (attached because it is too long):

[ids.md](https://github.com/user-attachments/files/22115625/ids.md)

jayaddison replied:
Thanks very much @gastmaier - that makes the problem and fix nice and clear.

I'm reading the comparison file at the moment - in particular I'm interested to find whether _any_ of the `before` elements included a `/` delimiter -- I haven't found any so far.  If there are none, then that would completely resolve my concern about breaking any existing hyperlinks containing that character.

Do you have any thoughts about whether we should _always_ include the complete document path prefix?  Or whether, for example, it could be omitted for unambiguous/unique IDs?

gastmaier replied:
I would patch all at the moment, it makes sense to me to store the lost docname information in the id itself, and it is clearer to debug.

For the toctree, before the pr, it would already generate in the format `document-<doc>#<id>`, so this would need to be assessed as well. That's what https://github.com/sphinx-doc/sphinx/pull/13717 tried to fix, only to uncover the collision issue.

And there are so many `visit_*` elements that needs to be patched to handle every corner case, that uniforming into a single format early on (after SphinxPostTransform, before other singlehtml patches) seems to be the only reliable approach.

The latex builder does patch at the `visit_*` elements with the `sphinx/writers/latex.py#hypertarget[withdoc=True]` method, but I don't see that working with html since it is straight up more convoluted since each visit would require some `if builder.name is 'singlehtml'`.

jayaddison replied:
Thanks @gastmaier.  I agree.  Also: in my opinion we should highlight this as a hyperlink-breaking change -- for `singlehtml` builds -- in the changelog notes.

One remaining concern I have: the `document-` prefix on every target seems fairly verbose.  I am wondering whether we could restore the path delimiter (`/`) -- and maybe even add a root prefix (`/{docname}/#{refid}`) -- to reduce the risk of duplication against other anchors declared by projects.

gastmaier replied:
I can give a try, but I need to know exactly the format we want to achieve, as I understood it would be , in the html:

```
href="#/path/to/doc/#anchor_name"
```
?

jayaddison replied:
That's the same format that I had in mind too, yep :+1:

gastmaier committed:
[singlehtml] Reformat fignum and secnum tuple

Format as ``/docname/#id`` to match other parts.

jayaddison commented on CHANGES.rst:
```diff
@@ -66,6 +66,11 @@ Features added
   Patch by Jean-François B.
+* #13739: singlehtml builder: append the docname to ids with format
+  ``/<docname>/#<id>``, to ensure uniqueness. For example, ``id3`` becomes
+  ``/path/to/doc/#id3``. This is a breaking change since it alters the format
```
```suggestion
  ``/path/to/doc/#id3``. This will break existing hyperlinks to `singlehtml` HTML documents since it alters the format
```

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -88,29 +88,47 @@ def _get_local_toctree(
+    def prefix_ids_with_docname(self, tree: nodes.document) -> None:
+        # Append docname to refids and ids using format document-<docname>#<id>.
+        # Compensates for loss of the pathname section of the href, that
+        # ensures uniqueness in the html builder.
```
What do you think about adjusting the name of this method to `ensure_fully_qualified_refids`, or similar?

Note: when I suggest the terminology `fully-qualified` there, I'm borrowing it as a name from DNS: https://en.wikipedia.org/wiki/Fully_qualified_domain_name

gastmaier replied:
Will do, I see in the codebase similar `ensure_` methods that alter the doc without return.
I was going to suggest `get_fully_qualified_refids`, but it is not a get method so it would be misleading.

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -88,29 +88,47 @@ def _get_local_toctree(
+    def ensure_fully_qualified_refids(self, tree: nodes.document) -> None:
+        # Append docname to refids and ids using format document-<docname>#<id>.
+        # Compensates for loss of the pathname section of the href, that
+        # ensures uniqueness in the html builder.
```
Thanks for the method renaming!  Please also replace the comment with a [docstring](https://docs.python.org/3/tutorial/stdlib.html#quality-control).

gastmaier replied:
Done

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -121,8 +139,9 @@ def assemble_toc_fignumbers(
+        # Note: current Sphinx patches refid with docname to avoid confliction
+        #       in singlehtml mode.
+        #       To match the patch, it replaces key of secnumbers to
         #       tuple of docname and refid.
```
Nitpick: I would suggest not modifying this comment with this PR (e.g. reverting it to the original comment in this branch).  Although the format and the way the `docname` is added is changed slightly, the comment continues to communicate the purpose of the replacement (in my opinion).

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -88,13 +88,31 @@ def _get_local_toctree(
+                continue
+            env = doc.settings.env
+            if 'refid' in node or 'ids' in node:
+                docname = env.path2doc(doc['source'])
```
Is there any situation where `node.document` can be empty?  If not, then we can simplify:

```suggestion
            if 'refid' in node or 'ids' in node:
                docname = self.env.path2doc(node.document['source'])
```

gastmaier replied:
mypy complains
```
sphinx/builders/singlehtml.py:99:36: error: Value of type "document | None" is not indexable  [index]
```
I tried the NodeMatcher.findall but got the same result. 
With cast 
```
doc = cast(nodes.document, node.document)
```
I get
```
sphinx/builders/singlehtml.py:97:19: error: Redundant cast to "document"  [redundant-cast]
```

```assert node.document is not None``` works

jayaddison replied:
Nice, thank you :+1:

jayaddison commented on sphinx/builders/singlehtml.py:
```diff
@@ -88,13 +88,31 @@ def _get_local_toctree(
 
+    def ensure_fully_qualified_refids(self, tree: nodes.document) -> None:
+        """Append docname to refids and ids using format
+        document-<docname>#<id>. Compensates for loss of the pathname section
```
One more ID-format description to update here.

jayaddison reviewed (approved):
Looks great to me - thank you, @gastmaier!

Note that I am not a core contributor here, so my review is not binding and I cannot merge the PR - even so, it looks good to me.

cc @sphinx-doc/developers (does this GitHub team still exist?)
